controllers/storageclassclaim_controller.go

```diff
@@ -0,0 +1,484 @@
+// +kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
+
+func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	enqueueStorageConsumerRequest := handler.EnqueueRequestsFromMapFunc(
+		func(obj client.Object) []reconcile.Request {
+			annotations := obj.GetAnnotations()
+			if annotation, found := annotations[StorageClassClaimAnnotation]; found {
+				parts := strings.Split(annotation, "/")
+				return []reconcile.Request{{
+					NamespacedName: types.NamespacedName{
+						Namespace: parts[0],
+						Name:      parts[1],
+					},
+				}}
+			}
+			return []reconcile.Request{}
+		})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.StorageClassClaim{}, builder.WithPredicates(
+			predicate.GenerationChangedPredicate{},
+		)).
+		Watches(&source.Kind{Type: &storagev1.StorageClass{}}, enqueueStorageConsumerRequest).
+		Complete(r)
+}
+
+func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log = ctrllog.FromContext(ctx, "StorageClassClaim", request)
+	r.ctx = ctrllog.IntoContext(ctx, r.log)
+	r.log.Info("Reconciling StorageClassClaim.")
+
+	// Fetch the StorageClassClaim instance
+	r.storageClassClaim = &v1alpha1.StorageClassClaim{}
+	r.storageClassClaim.Name = request.Name
+	r.storageClassClaim.Namespace = request.Namespace
+
+	if err := r.get(r.storageClassClaim); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.Info("StorageClassClaim resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClassClaim.")
+		return reconcile.Result{}, err
+	}
+
+	r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimInitializing
+
+	ocsClientList := &v1alpha1.OcsClientList{}
+	if err := r.list(ocsClientList, client.InNamespace(r.OperatorNamespace)); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch l := len(ocsClientList.Items); {
+	case l == 0:
+		return reconcile.Result{}, fmt.Errorf("no OcsClient found")
+	case l != 1:
+		return reconcile.Result{}, fmt.Errorf("multiple OcsClient found")
+	}
+	r.ocsClient = &ocsClientList.Items[0]
+
+	var result reconcile.Result
+	var reconcileError error
+
+	// StorageCluster checks for required fields.
+	switch ocsClient := r.ocsClient; {
+	case ocsClient.Status.ConsumerID == "":
+		return reconcile.Result{}, fmt.Errorf("no external storage consumer id found on the " +
+			"OcsClient status, cannot determine mode")
+	case ocsClient.Spec.StorageProviderEndpoint == "":
+		return reconcile.Result{}, fmt.Errorf("no external storage provider endpoint found on the " +
+			"OcsClient spec, cannot determine mode")
+	}
+
+	result, reconcileError = r.reconcileConsumerPhases()
```

Comment:
As we dont have both provider and consumer in storageclassclaim anymore rename `reconcileConsumerPhases` to `reconcilePhases`?